SwiftVoiceQuickstart/ViewController.swift

```diff
@@ -116,111 +113,160 @@ class ViewController: UIViewController, TVONotificationDelegate, TVOCallDelegate
-        if let call = self.activeCall {
-            self.userInitiatedDisconnect = true
-            performEndCallAction(uuid: call.uuid)
-            self.toggleUIState(isEnabled: false, showCallControl: false)
-        } else {
-            let uuid = UUID()
-            let handle = "Voice Bot"
+        guard activeCall == nil else {
+            userInitiatedDisconnect = true
+            performEndCallAction(uuid: activeCall!.uuid)
+            toggleUIState(isEnabled: false, showCallControl: false)
             
-            self.checkRecordPermission { (permissionGranted) in
-                if (!permissionGranted) {
-                    let alertController: UIAlertController = UIAlertController(title: "Voice Quick Start",
-                                                                               message: "Microphone permission not granted",
-                                                                               preferredStyle: .alert)
-                    
-                    let continueWithMic: UIAlertAction = UIAlertAction(title: "Continue without microphone",
-                                                                       style: .default,
-                                                                       handler: { (action) in
-                        self.performStartCallAction(uuid: uuid, handle: handle)
-                    })
-                    alertController.addAction(continueWithMic)
-                    
-                    let goToSettings: UIAlertAction = UIAlertAction(title: "Settings",
-                                                                    style: .default,
-                                                                    handler: { (action) in
-                        UIApplication.shared.open(URL(string: UIApplicationOpenSettingsURLString)!,
-                                                  options: [UIApplicationOpenURLOptionUniversalLinksOnly: false],
-                                                  completionHandler: nil)
-                    })
-                    alertController.addAction(goToSettings)
-                    
-                    let cancel: UIAlertAction = UIAlertAction(title: "Cancel",
-                                                              style: .cancel,
-                                                              handler: { (action) in
-                        self.toggleUIState(isEnabled: true, showCallControl: false)
-                        self.stopSpin()
-                    })
-                    alertController.addAction(cancel)
-                    
-                    self.present(alertController, animated: true, completion: nil)
-                } else {
-                    self.performStartCallAction(uuid: uuid, handle: handle)
-                }
+            return
+        }
+        
+        let uuid = UUID()
+        let handle = "Voice Bot"
+        
+        checkRecordPermission { permissionGranted in
```

Comment:
Looks like this is async so probably need `[weak self]` and then `guard let self = self else { return }` to be completely safe.

Reply:
When there is a significant amount of code like this in a closure it can lead to a lot of `self.` which requires developer to be careful. So sometimes it makes sense to create a new function like `showMicAccessRequest` and call it from the closure. That way there is probably only one `self.` so less to be careful about. Just a strategy to be aware of.

Reply:
I now pulled out the `UIAlertController` code to its own method and call it here, as you suggested.

I also used `[weak self]` in the closures to avoid retain cycles. Because we only call methods from this closure and do not access any non-optional variables, pure optional-chaining (i.e. `self?.myMethod()`) instead of `guard let self = self else { return }` should be sufficient for the moment (as `self?.myMethod()` simply does nothing, when *somehow* `self` becomes nil 😄 ).

Reply:
Exactly very nice!